Helper/Order.php

```diff
@@ -2293,4 +2310,59 @@ public function getOrderStoreIdByDisplayId($displayId)
 
         return ($order && $order->getStoreId()) ? $order->getStoreId() : null;
     }
+
+    /**
+     * Cancels an order while allowing its quote to be re-used, used for failed_payment webhook
+     *
+     * @param string $displayId
+     * @param string $immutableQuoteId
+     * @throws \Magento\Framework\Exception\AlreadyExistsException
+     */
+    public function cancelFailedPaymentOrder($displayId, $immutableQuoteId)
+    {
+        $order = $this->getExistingOrder($displayId);
+        if (!$order || $order->getState() == OrderModel::STATE_CANCELED) {
+            return;
+        }
+        if ($order->getState() !== OrderModel::STATE_PENDING_PAYMENT) {
+            throw new BoltException(
+                __(
+                    'Order Delete Error. Order is in invalid state. Order #: %1 State: %2 Immutable Quote ID: %3',
+                    $displayId,
+                    $order->getState(),
+                    $immutableQuoteId
+                ),
+                null,
+                CreateOrder::E_BOLT_GENERAL_ERROR
+            );
+        }
+        $this->orderManagement->cancel($order->getId());
+        /** reload order because {@see \Bolt\Boltpay\Helper\Cart::getOrderByIncrementId} */
+        $order = $this->orderRepository->get($order->getId());
+
+        $parentQuoteId = $order->getQuoteId();
+        $parentQuote = $this->cartHelper->getQuoteById($parentQuoteId);
+        $this->_eventManager->dispatch(
+            'sales_model_service_quote_submit_failure',
+            [
+                'order' => $order,
+                'quote' => $parentQuote
+            ]
+        );
+
+        // reactivate session quote - the condiotion excludes PPC quotes
+        if ($parentQuoteId != $immutableQuoteId) {
+            $this->cartHelper->quoteResourceSave($parentQuote->setIsActive(true));
+        }
+        // reset PPC quote checkout type so it can be treated as active
+        if ($parentQuote->getBoltCheckoutType() == CartHelper::BOLT_CHECKOUT_TYPE_PPC_COMPLETE) {
+            $parentQuote->setBoltCheckoutType(CartHelper::BOLT_CHECKOUT_TYPE_PPC);
+            $this->cartHelper->quoteResourceSave($parentQuote->setIsActive(false));
+        }
+        $this->eventsForThirdPartyModules->dispatchEvent("beforeDeleteOrder", $order);
```

Comment:
Why we call beforeDeleteOrder?

Reply:
It seems that we need to restores Amasty Giftcard balances, but I think we would better create another event for this, just separate from deleting order.

Reply:
It seems that we need to restores Amasty Giftcard balances, but I think we would better create another event for this, just separate from deleting order.

Reply:
@dumega correct https://github.com/BoltApp/bolt-magento2/blob/master/Model/EventsForThirdPartyModules.php#L61
I will do this in a separate PR.